src/lib/misc/roughtime/roughtime.cpp

```diff
@@ -0,0 +1,370 @@
+     const auto serverPublicKey = Botan::Ed25519_PublicKey(Botan::base64_decode(l.at("serverPublicKey").get<std::string>()));
+     const auto nonceOrBlind = vector_to_array<64>(Botan::unlock(Botan::base64_decode(l.at("nonceOrBlind").get<std::string>())));
+     const auto response = Botan::unlock(Botan::base64_decode(l.at("reply").get<std::string>()));
+     if(publicKeyType != "ed25519")
+        throw Not_Implemented("only ed25519 publicKeyType is implemented");
+     m_links.push_back({response, serverPublicKey, nonceOrBlind});
+     }
+  }
+std::vector<Response> Chain::responses() const
+  {
+  std::vector<Response> responses;
+  for(unsigned i = 0; i < m_links.size(); ++i)
+    {
+    const auto& l = m_links[i];
+    const auto nonce = i ? nonce_from_blind(m_links[i-1].response, l.nonce_or_blind) : l.nonce_or_blind;
+    const auto response = Response::from_bits(l.response, nonce);
+    if(!response.validate(l.public_key))
+      throw Invalid_Argument("invalid signature or public key");
+    responses.push_back(response);
+    }
+  return responses;
+  }
+void Chain::next_blind_and_nonce(RandomNumberGenerator& rng, std::array<uint8_t, 64>& nonce, std::array<uint8_t, 64>& blind) const
+  {
+    rng.randomize(blind.data(), blind.size());
+    if(!m_links.empty())
+      {
+      const auto& previous = m_links.back();
+      nonce = nonce_from_blind(previous.response, blind);
+      }
+    else
+      {
+      nonce = blind;
+      }
+  }
+void Chain::append(const std::vector<uint8_t>& response,
+                  const Ed25519_PublicKey& public_key,
+                  const std::array<uint8_t, 64>& nonce_or_blind,
+                  size_t max_chain_size)
+  {
+  if(max_chain_size <= 0)
+    throw Invalid_Argument("max chain size must be positive");
+
+  while(m_links.size() >= max_chain_size)
+    {
+    if(m_links.size() == 1)
+      {
+        const auto nonce = nonce_from_blind(m_links[0].response, nonce_or_blind); //we need to convert blind to nonce
+        m_links.clear();
+        m_links.push_back({response, public_key, nonce});
+        return;
+      }
+    if(m_links.size() >= 2)
+      {
+        m_links[1].nonce_or_blind = nonce_from_blind(m_links[0].response, m_links[1].nonce_or_blind); //we need to convert blind to nonce
+      }
+    m_links.erase(m_links.begin());
+    }
+  m_links.push_back({response, public_key, nonce_or_blind});
+  }
+
+void Chain::persist(DataSink_Stream& ds) const
+  {
+  nlohmann::json all;
+  for(const auto &link : m_links)
+    {
+    nlohmann::json l =
+      {
+      {"publicKeyType", "ed25519"},
+      {"serverPublicKey", Botan::base64_encode(link.public_key.get_public_key())},
+      {"nonceOrBlind", Botan::base64_encode(link.nonce_or_blind.data(), link.nonce_or_blind.size())},
+      {"reply", Botan::base64_encode(link.response)},
+      };
+    all["links"].push_back(l);
+    }
+  const auto dump = all.dump(2);
+  ds.write((const uint8_t*)dump.data(), dump.size());
```

Comment:
For char*<->uint8_t* casts, use `cast_char_ptr_to_uint8` and `cast_uint8_ptr_to_char` from mem_ops.h